addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,160 @@
+            target = self.env['mail.channel'].browse(message.res_id)
+        elif self.env.user._is_public() and guest:
+            target = guest
+        else:
+            target = self.env.user.partner_id
+        self.env['bus.bus']._sendmany([(target, 'mail.link.preview/insert', link_previews._link_preview_format())])
+
+    @api.model
+    def _create_link_preview(self, url, message_id):
+        if self._is_domain_throttled(url):
+            return self.env['mail.link.preview']
+        link_preview_data = self._get_link_preview_from_url(url)
+        if link_preview_data:
+            link_preview_data['message_id'] = message_id
+            return self.sudo().create(link_preview_data)
+        return self.env['mail.link.preview']
+
+    def _delete_and_notify(self):
+        notifications = []
+        guest = self.env['mail.guest']._get_guest_from_context()
+        for link_preview in self:
+            if link_preview.message_id.model == 'mail.channel' and link_preview.message_id.res_id:
+                target = self.env['mail.channel'].browse(link_preview.message_id.res_id)
+            elif self.env.user._is_public() and guest:
+                target = guest
+            else:
+                target = self.env.user.partner_id
+            notifications.append((target, 'mail.link.preview/delete', {'id': link_preview.id}))
+        self.env['bus.bus']._sendmany(notifications)
+        self.unlink()
+
+    @api.model
+    def _is_link_preview_enabled(self):
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return link_preview_throttle > 0
+
+    @api.model
+    def _is_domain_throttled(self, url):
+        domain = urlparse(url).netloc
+        date_interval = fields.Datetime.to_string((datetime.now() - relativedelta(seconds=10)))
+        call_counter = self.search_count([
+            ('source_url', 'ilike', '%' + domain + '%'),
+            ('create_date', '>', date_interval),
+        ])
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return call_counter > link_preview_throttle
+
+    @api.model
+    def _get_link_preview_from_url(self, url):
+        try:
+            response = requests.head(url, timeout=3, headers=self.request_headers, allow_redirects=True)
+        except requests.exceptions.RequestException:
+            return False
+        if response.status_code != requests.codes.ok:
+            return False
+        image_mimetype = [
+            'image/bmp',
+            'image/gif',
+            'image/jpeg',
+            'image/png',
+            'image/tiff',
+            'image/x-icon',
+        ]
+        # Content-Type header can return a charset, but we just need the
+        # mimetype (eg: image/jpeg;charset=ISO-8859-1)
+        content_type = response.headers['Content-Type'].split(';')
+        if content_type[0] in image_mimetype:
```

Comment:
```suggestion
        image_mimetypes = (
            'image/jpeg',
            'image/png',
            'image/gif',
            'image/bmp',
            'image/tiff',
            'image/x-icon',
        )
        if content_type.startswith(image_mimetypes):
```

Reply:
I still need the split to extract the mime type. But if you really want to use `startwidth`, I don't mind.